compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULinkExecutables.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/LLVMGPU/Passes.h"
+#include "iree/compiler/Codegen/Utils/LinkingUtils.h"
+#include "iree/compiler/Utils/ModuleUtils.h"
+#include "llvm/Support/FormatVariadic.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_LLVMGPULINKEXECUTABLESPASS
+#include "iree/compiler/Codegen/LLVMGPU/Passes.h.inc"
+
+namespace {
+
+static bool isAddressSpacePrivate(uint32_t addressSpace) {
+  return addressSpace == /*local*/ 3 || addressSpace == /*private*/ 5;
+}
+
+static SymbolTable::Visibility
+convertLinkageToVisibility(LLVM::Linkage linkage) {
+  switch (linkage) {
+  case LLVM::Linkage::Private:
+    return SymbolTable::Visibility::Private;
+  case LLVM::Linkage::External:
+    return SymbolTable::Visibility::Public;
+  default:
+    return SymbolTable::Visibility::Public;
+  }
+}
+
+// Returns true if we are allowed to rename |op| as part of merging.
+// The LLVMGPU lowering is super careful about assigning linkage so we err on
+// the side of renaming (as 100% of usage today does not reference external
+// things).
+static bool allowRenamingPrivateLLVMSymbols(Operation *op) {
+  if (auto globalOp = dyn_cast<LLVM::GlobalOp>(op)) {
+    if (isAddressSpacePrivate(globalOp.getAddrSpace())) {
+      return true;
+    }
+    return convertLinkageToVisibility(globalOp.getLinkage()) ==
+           SymbolTable::Visibility::Private;
+  } else if (auto funcOp = dyn_cast<LLVM::LLVMFuncOp>(op)) {
+    return convertLinkageToVisibility(funcOp.getLinkage()) ==
+           SymbolTable::Visibility::Private;
+  }
+  return SymbolTable::getSymbolVisibility(op) ==
+         SymbolTable::Visibility::Private;
+}
+
+struct LLVMGPULinkExecutablesPass
+    : public impl::LLVMGPULinkExecutablesPassBase<LLVMGPULinkExecutablesPass> {
+  LLVMGPULinkExecutablesPass() = default;
+  void runOnOperation() override {
+    auto moduleOp = getOperation();
+    auto moduleBuilder = OpBuilder::atBlockBegin(moduleOp.getBody());
+
+    auto sourceExecutableOps =
+        llvm::to_vector<8>(moduleOp.getOps<IREE::HAL::ExecutableOp>());
+    if (sourceExecutableOps.size() <= 1)
+      return;
+
+    // Guess a module name, if needed, to make the output files readable.
+    auto moduleName = guessModuleName(moduleOp, "llvm_module");
+
+    // Create our new "linked" hal.executable.
+    std::string linkedExecutableName =
+        llvm::formatv("{0}_linked_{1}", moduleName, "llvm_gpu");
+    auto linkedExecutableOp = moduleBuilder.create<IREE::HAL::ExecutableOp>(
+        moduleOp.getLoc(), linkedExecutableName);
```

Comment:
What happens if you compile for two different llvm_gpu backends (e.g. CUDA + ROCm)? Do both linked executables get the same name?

Reply:
that is a fantastic question. damnit.

Reply:
It does exactly what one would expect, if one expected something. I don't expect anything. It does work, though. It may lead to some odd naming but CUDA + ROCM together will end up in the same executable as the linking pass doesn't care about targets (it just links all targets to appropriate variants). That's a flaw in the existing LLVMCPU pass too. I've scoped both passes to just the target backend they are linking - it's not great, but it will probably hold us until LLVM linking is done properly.